[숫자 야구 게임] 최인호 미션 제출합니다.

inhooo00 opened the pull request:
### 어려웠던 점
MVC모델 이해하기가 굉장히 어려웠다. spring boot보다 어려웠다..
### 궁금한 점
숫자가 맞는지 판단하는 알고리즘을 MVC 모델에서 어떤 패키지로 빼야하는지 궁금합니다.
dto 패키지를 만들었는데, 사용해도 되는지 궁금합니다.
서로 의논해가면서 이 방법은 어떨까 저 방법은 어떨까 라는 느낌으로 같이 코드를 구현했는데, 페어프로그래밍이 이런 것인지 궁금합니다.

hangillee committed:
Update README.md

hangillee committed:
Update README.md

YehyeokBang committed:
docs: 구현할 기능 목록 추가

- 입력, 출력, 기능으로 분류하여 과제 진행합니다.

inhooo00 committed:
docs: 구현할 입력 기준 추가

inhooo00 committed:
feat: 게임 시작 메세지 출력 기능 구현

YehyeokBang committed:
feat: 숫자 입력 기능 구현

- 숫자 입력 메시지를 출력합니다.
- 기준에 맞지 않는 입력을 받은 경우 예외가 발생합니다.
- 숫자를 가진 객체 내부에서 값을 검증합니다.

inhooo00 committed:
feat: 컴퓨터 난수 생성 기능 구현

- 숫자 관리를 문자열 기반에서 리스트 기반으로 변경합니다.

YehyeokBang committed:
feat: 게임 결과 출력 기능 구현

YehyeokBang committed:
fix: 입력받은 값이 아스키 값으로 반환되는 로직 수정

YehyeokBang committed:
feat: 야구 게임 진행 기능 구현

- 재시작 로직은 아직 미구현 상태입니다.

inhooo00 committed:
feat: 재시작 기능 구현

- 1을 입력하면 재시작합니다.

YehyeokBang committed:
docs: 1차 리뷰 후 개선 방향 추가

jiyoungmerong commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+
+            outputView.printResultMessage(result.strike(), result.ball());
+
+            if (result.strike() == 3) {
```
지금은 간단한 코드이지만! 나중에 인호님이 작성하신 코드를 누군가 유지보수 해야한다면, 여기서 나타내는 3이 어떤 의미인지 모를 수도 있을 것 같아요.
또한, 만약 값이 3에서 4로 변경된다면 3이 쓰인 모든 자리를 4로 변경해야 하는 불편함이 있을 것 같아요!

- 상수란 무엇일까요?
잘 모르겠다면 이 블로그를 참고해보는 것도 좋을 것 같아요. https://doing7.tistory.com/146

jiyoungmerong commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+                    computer.generateRandomNumber();
+                    continue;
+                }
+                break;
```
만약 조건이 늘어난다면 어떻게 될까요?
조건에 맞지 않을 경우 바로 함수를 빠져나갈 수 있도록 코드를 수정해보는 건 어떨까요?
- Early return pattern이란 무엇일까요?
- https://woonys.tistory.com/209 이 블로그에서 정말 잘 정리해둔 것 같아요!

jiyoungmerong commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+
+    }
+
+    private Result compare(List<Integer> userNumber, List<Integer> computerNumber) {
```
무엇과 무엇을 비교하는 메소드인지 메소드 이름에서부터 의도를 파악할 수 있도록 수정해보는 건 어떨까요?

jiyoungmerong commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,73 @@
+        set.addAll(computerNumber);
+
+        return 6 - set.size();
+    }
```
위 메소드는 비지니스 로직으로 판단되어 컨트롤러보단 모델 또는 도메인에 넣으면 좋을 것 같다고 생각하는데, 인호님 생각은 어떠신지 궁금해요!
+ 궁금하다고 적어주셨는데, MVC에서 Model, View, Controller가 각각 어떤 역할을 하는지 생각해보면 좋을 것 같아요!

그리고 countStrike 메소드의 반환타입은 Integer인데, 메소드 내의 strike는 int로 설정되어 있네요! 
이렇게 설정한 이유가 있을까요? 해당 부분에서 autoboxing은 발생하지 않을까요?

마지막으로 countBall 메소드의 리턴값이 하드코딩인 것 같아요! 어떻게 수정하면 좋을까요?

jiyoungmerong commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,31 @@
+        generateRandomNumber();
+    }
+
+    public void generateRandomNumber() {
+        List<Integer> computer = new ArrayList<>();
+        while (computer.size() < 3) {
+            int randomNumber = Randoms.pickNumberInRange(1, 9);
+            if (!computer.contains(randomNumber)) {
+                computer.add(randomNumber);
+            }
+        }
+        this.number = computer;
+    }
```
메소드 이름은 랜덤한 숫자를 생성하는 것으로 파악되는데, 
실제 역할은 랜덤 숫자 생성 및 contains로 중복검사 후 리스트에 추가하는 역할을 하고 있는 것 같아요.
메소드 이름을 변경하는 것보단 각각의 메소드로 분리하는 건 어떨까요?

jiyoungmerong commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+        return number;
+    }
+
+    private void validation(String inputNumber) {
+        if (!isNaturalNumber(inputNumber)) {
+            throw new IllegalArgumentException("1부터 9까지의 자연수를 입력해주세요.");
+        }
+        if (!isThreeDigitNumber(inputNumber)) {
+            throw new IllegalArgumentException("3자리수를 입력해주세요.");
+        }
+        if (!isDuplicate(inputNumber)) {
+            throw new IllegalArgumentException("중복되지 않은 수를 입력해주세요.");
+        }
+    }
```
하나의 validation이 여러 유효성 검사를 진행하고 있네요!
검사를 수행하는 메소드를 각각 분리하여, 하나의 책임을 가질 수 있도록 수정하는 건 어떨까요?
- 하나의 메소드에 작성하는 것과 각각의 메소드로 분리하는 것은 어떤 차이가 있을까요?

jiyoungmerong commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+        }
+    }
+
+    private boolean isNaturalNumber(String inputNumber) {
+        for (char c : inputNumber.toCharArray()) {
+            if (c < '1' || c > '9') {
+                return false;
+            }
+        }
+
+        return true;
+    }
```
자연수인지 검사하는 메소드로 보여요. 
특정 패턴의 메소드를 포함하고 있는지 matches 메소드를 활용해보는 건 어떨까요?
- 정규표현식, 즉 정규식이 무엇일까요?

jiyoungmerong commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+    }
+
+    private boolean isThreeDigitNumber(String inputNumber) {
+        return inputNumber.length() == 3;
```
매개변수로 null값이 들어온다면 어떻게 될까요?

jiyoungmerong commented on src/main/java/baseball/model/Number.java:
```diff
@@ -0,0 +1,68 @@
+    private List<Integer> stringToList(String inputNumber) {
+        List<Integer> list = new ArrayList<>();
+
+        for (int i = 0; i < inputNumber.length(); i++) {
+            list.add(getNumericValue(inputNumber.charAt(i)));
+        }
```
기회가 된다면 Stream을 사용해보는 건 어떨까요!

jiyoungmerong commented on src/main/java/baseball/view/OutputView.java:
```diff
@@ -0,0 +1,27 @@
+        }
+        if (strike > 0 && ball == 0) {
+            sb.append(strike).append("스트라이크");
+            if (strike == 3) {
+                sb.append("\n3개의 숫자를 모두 맞히셨습니다! 게임 종료");
+            }
```
이 조건이 if (strike > 0 && ball == 0) 해당 조건 안에 있는 이유가 무엇일까요?!
더 간결하게 할 수 있는 방법이 있지 않을까 싶어요!

jiyoungmerong commented on src/test/java/baseball/ApplicationTest.java:
```diff
@@ -21,13 +21,37 @@ class ApplicationTest extends NsTest {
 
     @Test
-    void 예외_테스트() {
+    void 세자리가_아닌_숫자_입력_예외_테스트_1() {
+        assertSimpleTest(() ->
+                assertThatThrownBy(() -> runException("12"))
+                        .isInstanceOf(IllegalArgumentException.class)
+        );
+    }
+
+    @Test
+    void 세자리가_아닌_숫자_입력_예외_테스트_2() {
         assertSimpleTest(() ->
                 assertThatThrownBy(() -> runException("1234"))
                         .isInstanceOf(IllegalArgumentException.class)
         );
     }
 
```
예외 테스트를 1000번 진행해야한다고 하면, 하나의 메소드로 만드는 것도 좋겠죠?!
물론 간단한 코드에서는 위와 같이 작성해도 괜찮다고 생각합니다 ㅎㅎ

jiyoungmerong commented on src/main/java/baseball/view/InputView.java:
```diff
@@ -0,0 +1,24 @@
+
+    public Number getInputNumber() {
+        System.out.print("숫자를 입력해주세요 : ");
+        String input = Console.readLine();
```
Console.readLine()은 View에서 처리되어야 할까요?
View의 역할이 무엇일까요?

jiyoungmerong reviewed (commented):
1. 숫자가 맞는지 판단하는 알고리즘을 MVC 모델에서 어떤 패키지로 빼야하는지 궁금합니다.
=> 리뷰 달아놨습니다!
2. dto 패키지를 만들었는데, 사용해도 되는지 궁금합니다.
=> DTO 패키지 자체를 사용해도 되는 건지 여쭤보시는 걸까요??
3. 서로 의논해가면서 이 방법은 어떨까 저 방법은 어떨까 라는 느낌으로 같이 코드를 구현했는데, 페어프로그래밍이 이런 것인지 궁금합니다.
=> 애자일!! 페어 프로그래밍 멋져요!

inhooo00 committed:
feat: 1차 리뷰 후 개선 방향 추가

: 하드코딩 수정, Integer 통일

YehyeokBang committed:
refactor: 사용자가 입력한 숫자 관련 모델 수정

- 전달받은 String 객체가 null인지 확인하는 절차를 추가합니다.
- matches() 메서드와 정규식을 사용하여 자연수인지 확인합니다.
- Stream API를 사용하여 전달받은 값을 리스트로 변환합니다.
- 확인하는 메서드들을 정적 메서드로 변경합니다.

inhooo00 committed:
docs: 오탈자 수정

inhooo00 committed:
refactor: 코드 리팩토링

- 숫자 비교 로직을 모델로 이동합니다.
- 최대한 한 개의 책임을 가지도록 메서드를 분리합니다.

YehyeokBang committed:
refactor: 추가 리팩토링

- 매직 넘버, 문자열을 상수화합니다.
- 일부 조건문들을 early return으로 변경합니다.

inhooo00 commented:
# 1차 리뷰 후 개선 방향
- `View`의 수정에 `Model`이 영향을 받지 않도록 분리하기, 반대로도 마찬가지
- 가독성 향상하기 (`indent 줄이기` -> `early return`, `메서드 분리` 등)
- `Controller`의 역할 다시 생각해보기 
- 의미 있는 이름의 상수로 대체될 수 있는 수 또는 문자열 확인하기
- EOL 학습하기 
- generateRandomNumber 메서드 -> 각각의 메서드로 나누기
- validation 부분 -> 각각의 메서드로 나누기
- 자연수 확인하는 로직 정규식 써보자.. (시도했다가 실패했던 그거)
- printResultMessage에서 if문안에 if문을 넣은 이유 고민해보기
- view에서 Console.readLine()이 처리되는 게 맞는지 고민해보기
를 중점으로 생각하며 리팩토링해봤습니다!

jiyoungmerong commented on src/main/java/baseball/constant/ViewMessage.java:
```diff
@@ -0,0 +1,15 @@
+package baseball.constant;
+
+public class ViewMessage {
```
viewMessage면 view패키지로 이동시키는 건 어떨까요!?
만약 추후에 추가해야 할 상수들이 많아진다면, 열거형 enum 사용을 추천드려요!
- https://juyeop.tistory.com/42
위 블로그에서 enum의 장점에 대해 잘 설명해둔 것 같아요!

jiyoungmerong commented on src/main/java/baseball/controller/BaseballGameController.java:
```diff
@@ -0,0 +1,44 @@
+
+public class BaseballGameController {
+
+    InputView inputView = new InputView();
+    OutputView outputView = new OutputView();
```
InputView와 OutputView 객체를 직접 생성하는 것과 의존성 주입을 사용하는 것은 무슨 차이가 있을까요!

jiyoungmerong commented on src/main/java/baseball/model/Computer.java:
```diff
@@ -0,0 +1,64 @@
+        return new Result(strike, ball);
+    }
+
+    private int countStrike(List<Integer> userNumbers) {
+        return (int) IntStream.range(0, NUMBERS_SIZE)
+                .filter(i -> Objects.equals(this.numbers.get(i), userNumbers.get(i)))
+                .count();
+    }
+
+
+    private int countBall(List<Integer> userNumbers) {
+        return countDuplicateNumber(this.numbers, userNumbers);
+    }
+
+    // Set 자료구조를 사용해 중복되는 수의 개수를 반환하는 메서드
+    private int countDuplicateNumber(List<Integer> list1, List<Integer> list2) {
+        int totalSize = list1.size() + list2.size();
+
+        Set<Integer> set = new HashSet<>();
+        set.addAll(list1);
+        set.addAll(list2);
+
+        return totalSize - set.size();
```
잘 수정해주셨습니다!
규모가 커진다면 중복 개수 반환같이 자주 사용되는 메소드들은 별도의 유틸리티 클래스를 만들어주면,
유지보수성도 높아지고 가독성도 좋아진답니다!!

jiyoungmerong commented on src/main/java/baseball/model/Numbers.java:
```diff
@@ -0,0 +1,66 @@
+        return numbers;
+    }
+
+    private static void checkInputIsNull(String input) {
+        if (input == null) {
+            throw new IllegalStateException("잘못된 인자를 전달받았습니다.");
+        }
+    }
+
+    private static void checkThreeDigitNumber(String inputNumbers) {
+        if (inputNumbers.length() != 3) {
+            throw new IllegalArgumentException("3자리수를 입력해주세요.");
+        }
+    }
+
+    private static void checkNaturalNumber(String inputNumbers) {
+        if (!inputNumbers.matches("[1-9]+")) {
+            throw new IllegalArgumentException("1부터 9까지의 자연수를 입력해주세요.");
+        }
+    }
```
메소드명만으로 주석 없이 어떤 역할을 하는지 잘 나타내주셨네요!!
예외처리를 메소드로 분리한 점도 좋습니다.

jiyoungmerong commented on src/main/java/baseball/model/Numbers.java:
```diff
@@ -0,0 +1,66 @@
+            throw new IllegalArgumentException("중복되지 않는 수를 입력해주세요.");
+        }
+
+        // GC가 처리하도록 명시적으로 null로 설정
+        set = null;
```
명시적으로 null을 할당해주면 GC가 메모리 회수를 보다 빨리 하겠죠!
큰 객체나 컬렉션 다룰 때 좋습니다! 굳!

jiyoungmerong commented on src/main/java/baseball/model/Numbers.java:
```diff
@@ -0,0 +1,66 @@
+                .filter(Character::isDigit)
+                .mapToObj(c -> (Character.digit(c, 10)))
+                .collect(Collectors.toList());
+    }
```
Stream활용 좋아요! 이것도 유틸리티 클래스에 들어가면 좋을 것 같아요!

jiyoungmerong reviewed (commented):
작은 규모임에도 불구하고 GC를 고려하여 코드작성한 점 멋있습니다!
